[APM] Automatically create APM index pattern

ogupte opened the pull request:
Closes elastic/kibana#33119 by automatically creating the apm index pattern on server startup if one isn't found.

This is a continuation of https://github.com/elastic/kibana/pull/34338 until it's ready to be merged.

ogupte committed:
Closes #33119 by automatically creating the apm index pattern on server startup if one isn't found.

ogupte committed:
[APM] Move lib/saved_objects/client -> lib/helpers/saved_objects_client

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8506/)

elasticmachine commented:
Pinging @elastic/apm-ui

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9740/)

jasonrhodes commented on x-pack/plugins/apm/server/routes/services.ts:
```diff
@@ -61,4 +62,6 @@ export function initServicesApi(core: CoreSetup) {
     }
   });
+
+  ensureIndexPatternExists(server);
```
Just curious: why is this inside the `initServicesApi` function?

ogupte replied:
Since it's supposed to create the index on server startup, it's being called while the plugin initializes.

jasonrhodes replied:
Cool, we talked about this on a call but to re-cap: I think it makes sense to call this from inside the new Plugin file where the init api calls are made, and pass in `core` which has `core.server` on it, rather than calling it from within one of the existing init calls.

jasonrhodes commented on x-pack/plugins/apm/server/lib/index_pattern/index.ts:
```diff
@@ -0,0 +1,25 @@
+ */
+import { Server } from 'hapi';
+// @ts-ignore
+import { getSavedObjects } from '../../../../../../src/legacy/core_plugins/kibana/server/tutorials/apm/saved_objects/get_saved_objects';
```
Does this even need to be mentioned in the tutorial section anymore if we always auto-create these objects?

ogupte replied:
i think it's worth showing the step in the tutorial, it's possible that the index pattern fails to save on startup, or gets deleted, so it's good visual confirmation that the step has been completed.

jasonrhodes replied:
Makes sense. I think it might be better to have the code live in our plugin and have the tutorial reference it rather than our plugin referencing tutorial code, although I'm not sure if there are any open source implications to that?

jasonrhodes commented:
@formgeist I addressed all feedback, I think, except for the 50% width thing -- I set the table to "table-layout: auto" which made the columns stop sticking to 50%, but they got really weird and some were big and some were small and it looked awful. I can show you on a call if you like. :)

ogupte committed:
[APM] moved apm index pattern from legacy tutorial directory to apm/common & moved index creation call to top level plugin initialization

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10117/)

jasonrhodes commented:
Looks good, just need to verify if moving the "get saved objects" method into our plugin code breaks the OSS APM build ... not sure if that's what the broken CI build is related to or not. If that causes too many problems, maybe we just keep a copy of the method in both places?

ogupte committed:
[APM] moved the apm index_pattern.json back to the core kibana side of
things so that it doesn't break the oss only build by requireing a
non-existent module.

ogupte committed:
[APM] fixed wrong path import for index_pattern

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10481/)
